Simplify `TypeFreshener` methods.

nnethercote opened the pull request:
`freshen_{ty,const}` take a `Result` and just do a fold if the input is `Ok`. It's simpler to do those folds at the call site, and only call `freshen_{ty,const}` in the `Err` case. That way we can also avoid useless fold operations on the results of `new_{int,uint,float}`.

Also, make some `bug!` calls more concise.

r? @lcnr

nnethercote committed:
Simplify `TypeFreshener` methods.

`freshen_{ty,const}` take a `Result` and just do a fold if the input is
`Ok`. It's simpler to do those folds at the call site, and only call
`freshen_{ty,const}` in the `Err` case. That way we can also avoid
useless fold operations on the results of `new_{int,uint,float}`.

Also, make some `bug!` calls more concise.

nnethercote committed:
Remove `InferCtxt::freshener`.

I have always found this confusingly named, because it creates a new
freshener rather than returning an existing one. We can remove it and
just use `TypeFreshener::new()` at the two call sites, avoiding this
confusion.

nnethercote committed:
Remove `InferCtxt::freshen`.

Sometimes we freshen using a new `TypeFreshener`, and sometimes we
freshen with an existing `TypeFreshener`. For the former we have the
method `InferCtxt::freshen`. For the latter we just call `fold_with`.
This asymmetry has been confusing to me.

This commit removes `InferCtxt::freshen` so that all the freshening
sites consistently use `fold_with` and it's obvious if each one is using
a new or existing `TypeFreshener`.

lcnr reviewed (commented):
r=me, but would love for u to look into a related refactor

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150239/f41f40408d719aa9ae0c6bfa17619d8f3f9e5b99..0d4649da130e656570c5728ef840bac9e0d6e601)*

nnethercote committed:
`s/fresh_trait_ref/fresh_trait_pred`.

Because `fresh_trait_pred` is the name of the field/argument. The `_ref`
suffix appears to be a typo, or left over from earlier versions of the
code.

nnethercote committed:
Disallow freshening an already-freshened type/const.

It never happens in practice.

nnethercote committed:
Remove `Option` from `fold_infer_ty` return type.

It's no longer needed.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

lcnr commented:
@bors try

rust-timer commented:
Finished benchmarking commit (d1d7c443a2cb57b25b43cf0031489c133be86cff): [comparison URL](https://perf.rust-lang.org/compare.html?start=99ff3fbb86658b427f5dd7daaae8db5626a63c26&end=d1d7c443a2cb57b25b43cf0031489c133be86cff&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=99ff3fbb86658b427f5dd7daaae8db5626a63c26&end=d1d7c443a2cb57b25b43cf0031489c133be86cff&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=99ff3fbb86658b427f5dd7daaae8db5626a63c26&end=d1d7c443a2cb57b25b43cf0031489c133be86cff&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=99ff3fbb86658b427f5dd7daaae8db5626a63c26&end=d1d7c443a2cb57b25b43cf0031489c133be86cff&stat=cycles:u)
<details>
<summary>Results (primary 2.1%, secondary 2.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1% | [1.7%, 2.6%] | 4     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [2.1%, 2.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.1% | [1.7%, 2.6%] | 4     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=99ff3fbb86658b427f5dd7daaae8db5626a63c26&end=d1d7c443a2cb57b25b43cf0031489c133be86cff&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 482.582s -> 480.862s (-0.36%)
**Artifact size**: 390.35 MiB -> 390.37 MiB (0.00%)

lcnr commented:
@bors r+ rollup

bors commented:
:pushpin: Commit ec5e96e58b55e1161f69e6515bb3f070f41c57ec has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr ec5e96e58b55e1161f69e6515bb3f070f41c57ec -->
<!-- homu: {"type":"Approved","sha":"ec5e96e58b55e1161f69e6515bb3f070f41c57ec","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
